File: projeto/backend/nest/src/modules/domain/models/users/dto/create-user.dto.ts
import { IsArray, IsEmail, IsEnum, IsNotEmpty, IsString, IsStrongPassword, MaxLength } from "class-validator";
import { Permission } from "src/common/enums/permissao.enum";

export class CreateUserDto {
  @IsEmail()
  @IsNotEmpty()
  email: string;

  @IsString()
  @IsNotEmpty()
  @MaxLength(255)
  name: string;

  @IsString()
  @IsNotEmpty()
  @IsStrongPassword({
    minLength: 8,
    minLowercase: 1,
    minUppercase: 1,
    minNumbers: 1,
    minSymbols: 1,
  }, { message: 'Password must contain at least 8 characters, 1 uppercase letter, 1 lowercase letter, 1 number, and 1 symbol' })
  password: string;

  @IsString()
  @IsNotEmpty()
  confirmPassword: string;

  @IsNotEmpty()
  @IsArray()
  @IsEnum(Permission, { each: true, message: 'Permissions must be an array of valid Permission enum values' })
  permissions: Permission[];
}

File: projeto/backend/csharp_p2/src/Shared/DTOs/Users/CreateUserDto.cs
using System.ComponentModel.DataAnnotations;

namespace csharp_p2.src.Shared.DTOs.Users;

public class CreateUserDto {
  [Required(ErrorMessage = "Required field.")]
  [EmailAddress(ErrorMessage = "Email is not valid.")]
  public string Email { get; set; } = null!;

  [Required(ErrorMessage = "Required field.")]
  [MinLength(8, ErrorMessage = "Password must contain at least 8 characters.")]
  [RegularExpression(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[\W_]).+$",
          ErrorMessage = "Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number and 1 symbol.")]
  public string Password { get; set; } = null!;

  [Required(ErrorMessage = "Required field.")]
  [MinLength(8, ErrorMessage = "Password must contain at least 8 characters.")]
  [RegularExpression(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[\W_]).+$",
          ErrorMessage = "Password must contain at least 1 uppercase letter, 1 lowercase letter, 1 number and 1 symbol.")]
  public string ConfirmPassword { get; set; }

  [Required(ErrorMessage = "Required field.")]
  [StringLength(255, MinimumLength = 3, ErrorMessage = "Name must be between 3 and 255 characters.")]
  public string Name { get; set; }

  [Required(ErrorMessage = "Required field.")]
  public long? RoleId { get; set; }
}
